src/FSharp.Analyzers/DisposedBeforeAsyncRunAnalyzer.fs

```diff
@@ -0,0 +1,169 @@
+open FSharp.Compiler.Symbols
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.SyntaxTrivia
+open FSharp.Compiler.Text
+
+[<Literal>]
+let Code = "GRA-DISPBEFOREASYNC-001"
+
+let getType (checkFileResults : FSharpCheckFileResults) (sourceText : ISourceText) (synPat : SynPat) =
+    let lineText = sourceText.GetLineString (synPat.Range.EndLine - 1) // 0-based
+
+    match synPat with
+    | SynPat.LongIdent (longDotId = SynLongIdent (id = idents)) -> idents |> List.tryLast
+    | SynPat.Named (ident = SynIdent (ident = ident)) -> Some ident
+    | _ -> None
+    |> Option.bind (fun i ->
+        checkFileResults.GetSymbolUseAtLocation (synPat.Range.EndLine, synPat.Range.EndColumn, lineText, [ i.idText ])
+        |> Option.bind (fun symbolUse ->
+            match symbolUse.Symbol with
+            | :? FSharpMemberOrFunctionOrValue as mfv when mfv.IsFunction -> mfv.FullTypeSafe
+            | _ -> None
+        )
+    )
+
+let asyncOrTask (t : FSharpType) =
+    t.GenericArguments
+    |> Seq.tryLast
+    |> Option.map (fun t ->
+        if t.HasTypeDefinition then
+            match t.TypeDefinition.TryGetFullName () with
+            | Some fullName when fullName.StartsWith ("Microsoft.FSharp.Control.FSharpAsync", StringComparison.Ordinal) ->
+                Some "async"
+            | Some fullName when fullName.StartsWith ("System.Threading.Tasks.Task", StringComparison.Ordinal) ->
+                Some "task"
+            | _ -> None
+        else
+            None
+    )
+    |> Option.flatten
+
+let pathContainsAsyncOrTaskReturningFunc
+    (checkFileResults : FSharpCheckFileResults)
+    (sourceText : ISourceText)
+    (path : SyntaxVisitorPath)
+    =
+    path
+    |> List.tryPick (fun node ->
+        match node with
+        | SyntaxNode.SynBinding (SynBinding (headPat = headPat)) ->
+            getType checkFileResults sourceText headPat |> Option.map asyncOrTask
+        | _ -> None
+    )
+    |> Option.flatten
+
+let pathContainsComputationExpr (path : SyntaxVisitorPath) =
+    path
+    |> List.exists (
+        function
+        | SyntaxNode.SynExpr (SynExpr.ComputationExpr _) -> true
+        | _ -> false
+    )
+
+[<Literal>]
+let SwitchOffAsyncComment = "disposed before returned async runs"
+
+[<Literal>]
+let SwitchOffTaskComment = "disposed before returned task runs"
+
+let collectUses (sourceText : ISourceText) (ast : ParsedInput) (checkFileResults : FSharpCheckFileResults) =
+    let comments =
+        match ast with
+        | ParsedInput.ImplFile parsedImplFileInput -> parsedImplFileInput.Trivia.CodeComments
+        | _ -> []
+
+    let isSwitchedOffPerComment (range : Range) =
+        comments
+        |> List.exists (fun c ->
+            match c with
+            | CommentTrivia.LineComment r ->
+                if r.StartLine <> range.StartLine - 1 then
+                    false
+                else
+                    let lineOfComment = sourceText.GetLineString (r.StartLine - 1) // 0-based
```

Comment:
Nit: you should try the `deindented` else branch here. It grows on you.

Reply:
Nah man, I have my limits ;)